proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1775 @@
+a relatively minimal change in the Cabal codebase.  
+Many Cabal contributors share a long term goal to move the Cabal design towards
+one based on a build graph with fine-grained dependencies. From this perspective,
+the critique was that the initial proposal was too conservative a change, and that
+we should take this opportunity of making a significant API change to establish a
+new API that would not hold back the move towards finer-grained dependencies.
+Another critique was that the original `UserHooks` design was somewhat ad-hoc,
+since it used both monolithic hooks and hooked pre-processors to provide
+finer-grained dependencies for a modest subset of use cases.  
+On the other hand, there is a very large design space for finer-grained
+dependencies, and so picking a point in the design space is not simple.
+Another disadvantage is that it will of course be more work for package authors
+to port their existing `Setup.hs` scripts, which currently use monolithic hooks.
+
+The proposed design for pre-build hooks tries to balance these trade-offs.
+Instead of an ad-hoc combination of monolithic hooks and hooked pre-processors,
+we use a single general system of rules, but we take a relatively conservative
+approach to the expressive power of the rules. In particular, the style of the
+rules is relatively low level. For example it does not include "rule patterns"
+such as generating a `*.hs` from a `*.y`. Instead, each rule specifies the
+individual files involved as inputs and outputs. It should nevertheless be
+possible to build higher level patterns on top, using Haskell's usual powers of
+abstraction to generate the lower level rules. Crucially, the design allows the
+rules to be used across an IPC interface, which is necessary for build tools
+like `cabal-install` or HLS to be able to interrogate and invoke them.  
+
+The full details of the design of pre-build hooks are provided in
+[§ Pre-build hooks](#pre-build-hooks).
+
+On top of pre-build hooks, we also propose a limited notion of post-build hooks,
+which accomodates package authors which need to perform an `IO` action in order
+to modify executables after they are built, as described in
+[§ Post-build hooks](#post-build-hooks).
+
+To summarise, we propose two different kinds of build hooks:
+
+```haskell
+-- | Build-time hooks.
+data BuildHooks
+  = BuildHooks
+  { preBuildComponentRules :: Maybe PreBuildComponentRules
+  , postBuildComponentHook :: Maybe PostBuildComponentHook }
+```
+
+Build hooks cannot change the configuration of the package.  
+There are deliberately no package-level build hooks, only component-level hooks.
+This avoids introducing unnecessary synchronisation points when multiple
+packages/components are being built in parallel.
+
+#### Post-build hooks
+
+Separately from the pre-build rules, we also propose to introduce post-build
+hooks. These cover a simple use case: namely to perform an IO action after an
+executable has been built.
```

Comment:
I think the observation that most post-build hooks are really for after an exectuable is built is enlightening.

There is a clear difference in the build process between building a library component and building an executable: they both build all the Haskell modules, but for an executable we additionally _link them together into an executable_.

So I wonder if we are missing a phase: the link phase. Most of the post-build hooks you describe are then actually post-_link_ hooks.